File: sdk-python/src/custd/admin_predictions.py
from __future__ import annotations

from typing import Any, TypedDict

from .client import AdminClient, quote_path


class PredictionDefinitionCreateRequest(TypedDict, total=False):
    definition_key: str
    display_name: str
    description: str


class PredictionDefinitionUpdateRequest(TypedDict, total=False):
    definition_key: str
    display_name: str
    description: str
    revision: int


class PredictionVersionPublishRequest(TypedDict):
    definition: dict[str, Any]
    created_by: str


class PredictionSignalSourceCreateRequest(TypedDict, total=False):
    source_key: str
    source_mode: str
    display_name: str
    description: str
    configuration: dict[str, Any]
    poll_interval_seconds: int


class PredictionActivateRequest(TypedDict):
    version_uuid: str


class PredictionRollbackRequest(TypedDict, total=False):
    version_uuid: str
    reason: str


class PredictionPauseRequest(TypedDict, total=False):
    reason: str


class PredictionRunNowRequest(TypedDict, total=False):
    worker_id: str


class PredictionAdminClient:
    def __init__(self, admin: AdminClient) -> None:
        self._admin = admin

    def list_definitions(
        self, company_slug: str, page_size: int | None = None, page_token: str | None = None
    ) -> dict[str, Any]:
        return self._admin.request("GET", self._collection("/definitions", company_slug, page_size, page_token))

    def get_definition(self, company_slug: str, definition_uuid: str) -> dict[str, Any]:
        return self._admin.request("GET", self._resource(f"/definitions/{definition_uuid}", company_slug))

    def create_definition(self, company_slug: str, body: PredictionDefinitionCreateRequest) -> dict[str, Any]:
        return self._admin.request("POST", self._collection("/definitions", company_slug), dict(body))

    def update_definition(
        self, company_slug: str, definition_uuid: str, body: PredictionDefinitionUpdateRequest
    ) -> dict[str, Any]:
        return self._admin.request(
            "PATCH", self._resource(f"/definitions/{definition_uuid}", company_slug), dict(body)
        )

    def get_version(self, company_slug: str, definition_uuid: str, version_uuid: str) -> dict[str, Any]:
        return self._admin.request(
            "GET", self._resource(f"/definitions/{definition_uuid}/versions/{version_uuid}", company_slug)
        )

    def publish_version(
        self, company_slug: str, definition_uuid: str, body: PredictionVersionPublishRequest
    ) -> dict[str, Any]:
        return self._admin.request(
            "POST", self._resource(f"/definitions/{definition_uuid}/publish", company_slug), dict(body)
        )

    def activate_version(
        self, company_slug: str, definition_uuid: str, body: PredictionActivateRequest
    ) -> dict[str, Any]:
        return self._admin.request(
            "POST", self._resource(f"/definitions/{definition_uuid}/activate", company_slug), dict(body)
        )

    def rollback_version(
        self, company_slug: str, definition_uuid: str, body: PredictionRollbackRequest
    ) -> dict[str, Any]:
        return self._admin.request(
            "POST", self._resource(f"/definitions/{definition_uuid}/rollback", company_slug), dict(body)
        )

    def pause_definition(
        self, company_slug: str, definition_uuid: str, body: PredictionPauseRequest | None = None
    ) -> None:
        self._admin.request(
            "POST", self._resource(f"/definitions/{definition_uuid}/pause", company_slug), dict(body or {})
        )

    def resume_definition(self, company_slug: str, definition_uuid: str) -> None:
        self._admin.request("POST", self._resource(f"/definitions/{definition_uuid}/resume", company_slug))

    def archive_definition(self, company_slug: str, definition_uuid: str) -> None:
        self._admin.request("POST", self._resource(f"/definitions/{definition_uuid}/archive", company_slug))

    def run_now(self, company_slug: str, definition_uuid: str, body: PredictionRunNowRequest | None = None) -> None:
        self._admin.request(
            "POST", self._resource(f"/definitions/{definition_uuid}/run-now", company_slug), dict(body or {})
        )

    def list_runs(self, company_slug: str, definition_uuid: str, page_size: int | None = None) -> list[dict[str, Any]]:
        return self._list(f"/definitions/{definition_uuid}/runs", company_slug, page_size)

    def list_outcomes(
        self, company_slug: str, definition_uuid: str, page_size: int | None = None
    ) -> list[dict[str, Any]]:
        return self._list(f"/definitions/{definition_uuid}/outcomes", company_slug, page_size)

    def get_evaluation(self, company_slug: str, definition_uuid: str) -> dict[str, Any]:
        return self._admin.request("GET", self._resource(f"/definitions/{definition_uuid}/evaluations", company_slug))

    def list_threshold_events(
        self, company_slug: str, definition_uuid: str, page_size: int | None = None
    ) -> list[dict[str, Any]]:
        return self._list(f"/definitions/{definition_uuid}/threshold-events", company_slug, page_size)

    def list_signal_sources(
        self, company_slug: str, page_size: int | None = None, page_token: str | None = None
    ) -> list[dict[str, Any]]:
        return self._list("/sources", company_slug, page_size, page_token)

    def get_signal_source(self, company_slug: str, source_uuid: str) -> dict[str, Any]:
        return self._admin.request("GET", self._resource(f"/sources/{source_uuid}", company_slug))

    def create_signal_source(self, company_slug: str, body: PredictionSignalSourceCreateRequest) -> dict[str, Any]:
        return self._admin.request("POST", self._collection("/sources", company_slug), dict(body))

    def activate_signal_source(self, company_slug: str, source_uuid: str) -> None:
        self._admin.request("POST", self._resource(f"/sources/{source_uuid}/activate", company_slug))

    def archive_signal_source(self, company_slug: str, source_uuid: str) -> None:
        self._admin.request("POST", self._resource(f"/sources/{source_uuid}/archive", company_slug))

    def _list(
        self, path: str, company_slug: str, page_size: int | None = None, page_token: str | None = None
    ) -> list[dict[str, Any]]:
        response = self._admin.request("GET", self._collection(path, company_slug, page_size, page_token))
        return list(response.get("items", []))

    @staticmethod
    def _collection(path: str, company_slug: str, page_size: int | None = None, page_token: str | None = None) -> str:
        query = [f"companySlug={quote_path(company_slug)}"]
        if page_size is not None:
            query.append(f"pageSize={page_size}")
        if page_token is not None:
            query.append(f"pageToken={quote_path(page_token)}")
        return f"/measurement/predictions{path}?{'&'.join(query)}"

    @classmethod
    def _resource(cls, path: str, company_slug: str) -> str:
        segments = path.split("/")
        encoded = "/".join(segment if index == 0 else quote_path(segment) for index, segment in enumerate(segments))
        return cls._collection(encoded, company_slug)

File: sdk-js/src/admin-predictions.ts
import type { RequestOptions } from "./index.js";

export type PredictionDefinition = {
  uuid: string;
  definition_key: string;
  display_name: string;
  description?: string;
  status: string;
  schedule_kind: string;
  default_horizon_seconds?: number;
  is_paused: boolean;
  created_at: string;
  updated_at: string;
};

export type PredictionDefinitionCreateRequest = {
  definition_key: string;
  display_name: string;
  description?: string;
};

export type PredictionDefinitionUpdateRequest = {
  definition_key?: string;
  display_name?: string;
  description?: string;
  revision: number;
};

export type PredictionDefinitionListResponse = {
  items: PredictionDefinition[];
  next_page_token?: string;
};

export type PredictionVersion = {
  uuid: string;
  version_uuid: string;
  version_number: number;
  version_status: string;
  definition_hash: string;
  definition: Record<string, unknown>;
  feature_count: number;
  source_count: number;
  created_by: string;
  created_at: string;
};

export type PredictionVersionPublishRequest = {
  definition: Record<string, unknown>;
  created_by: string;
};

export type PredictionActivateRequest = { version_uuid: string };
export type PredictionRollbackRequest = { version_uuid: string; reason?: string };
export type PredictionPauseRequest = { reason?: string };
export type PredictionRunNowRequest = { worker_id?: string };

export type PredictionSignalSource = {
  uuid: string;
  source_key: string;
  source_mode: string;
  display_name: string;
  description?: string;
  poll_interval_seconds?: number;
  source_status: string;
  is_paused: boolean;
  created_at: string;
  updated_at: string;
  last_succeeded_at?: string;
  consecutive_failed_count: number;
};

export type PredictionSignalSourceCreateRequest = {
  source_key: string;
  source_mode: string;
  display_name: string;
  description?: string;
  configuration: Record<string, unknown>;
  poll_interval_seconds?: number;
};

export type PredictionRunSummary = {
  run_uuid: string;
  as_of_at: string;
  horizon_end_at: string;
  output: number;
  baseline: number;
  override_applied: boolean;
  input_hash: string;
  engine_version: string;
  warning_count: number;
  generated_at: string;
  duration_milliseconds?: number;
};

export type PredictionOutcomeSummary = {
  run_uuid: string;
  resolution: string;
  outcome_kind: string;
  resolved_at?: string;
  observed_value: number;
  is_late_evidence: boolean;
};

export type PredictionEvaluationSummary = {
  window_start_at: string;
  window_end_at: string;
  resolved_run_count: number;
  positive_count: number;
  negative_count: number;
  brier_score?: number;
  log_loss?: number;
  sparse_bucket_count: number;
};

export type PredictionThresholdEvent = {
  version_uuid: string;
  hysteresis_state: string;
  direction: string;
  observed_value: number;
  threshold: number;
  trigger_run_uuid: string;
  dedup_key: string;
  emitted_at: string;
};

type AdminRequester = <T>(method: string, path: string, body?: unknown, options?: RequestOptions) => Promise<T>;

export class PredictionAdminClient {
  constructor(private readonly request: AdminRequester) {}

  listDefinitions(companySlug: string, pageSize?: number, pageToken?: string, options?: RequestOptions) {
    return this.request<PredictionDefinitionListResponse>(
      "GET",
      collectionPath("/definitions", companySlug, pageSize, pageToken),
      undefined,
      options,
    );
  }

  getDefinition(companySlug: string, definitionUuid: string, options?: RequestOptions) {
    return this.request<PredictionDefinition>(
      "GET",
      resourcePath(`/definitions/${definitionUuid}`, companySlug),
      undefined,
      options,
    );
  }

  createDefinition(companySlug: string, body: PredictionDefinitionCreateRequest, options?: RequestOptions) {
    return this.request<PredictionDefinition>("POST", collectionPath("/definitions", companySlug), body, options);
  }

  updateDefinition(
    companySlug: string,
    definitionUuid: string,
    body: PredictionDefinitionUpdateRequest,
    options?: RequestOptions,
  ) {
    return this.request<PredictionDefinition>(
      "PATCH",
      resourcePath(`/definitions/${definitionUuid}`, companySlug),
      body,
      options,
    );
  }

  getVersion(companySlug: string, definitionUuid: string, versionUuid: string, options?: RequestOptions) {
    return this.request<PredictionVersion>(
      "GET",
      resourcePath(`/definitions/${definitionUuid}/versions/${versionUuid}`, companySlug),
      undefined,
      options,
    );
  }

  publishVersion(
    companySlug: string,
    definitionUuid: string,
    body: PredictionVersionPublishRequest,
    options?: RequestOptions,
  ) {
    return this.request<PredictionVersion>(
      "POST",
      resourcePath(`/definitions/${definitionUuid}/publish`, companySlug),
      body,
      options,
    );
  }

  activateVersion(
    companySlug: string,
    definitionUuid: string,
    body: PredictionActivateRequest,
    options?: RequestOptions,
  ) {
    return this.request<PredictionVersion>(
      "POST",
      resourcePath(`/definitions/${definitionUuid}/activate`, companySlug),
      body,
      options,
    );
  }

  rollbackVersion(
    companySlug: string,
    definitionUuid: string,
    body: PredictionRollbackRequest,
    options?: RequestOptions,
  ) {
    return this.request<PredictionVersion>(
      "POST",
      resourcePath(`/definitions/${definitionUuid}/rollback`, companySlug),
      body,
      options,
    );
  }

  pauseDefinition(
    companySlug: string,
    definitionUuid: string,
    body: PredictionPauseRequest = {},
    options?: RequestOptions,
  ) {
    return this.request<void>("POST", resourcePath(`/definitions/${definitionUuid}/pause`, companySlug), body, options);
  }

  resumeDefinition(companySlug: string, definitionUuid: string, options?: RequestOptions) {
    return this.request<void>(
      "POST",
      resourcePath(`/definitions/${definitionUuid}/resume`, companySlug),
      undefined,
      options,
    );
  }

  archiveDefinition(companySlug: string, definitionUuid: string, options?: RequestOptions) {
    return this.request<void>(
      "POST",
      resourcePath(`/definitions/${definitionUuid}/archive`, companySlug),
      undefined,
      options,
    );
  }

  runNow(companySlug: string, definitionUuid: string, body: PredictionRunNowRequest = {}, options?: RequestOptions) {
    return this.request<void>(
      "POST",
      resourcePath(`/definitions/${definitionUuid}/run-now`, companySlug),
      body,
      options,
    );
  }

  listRuns(companySlug: string, definitionUuid: string, pageSize?: number, options?: RequestOptions) {
    return this.request<PredictionRunSummary[]>(
      "GET",
      collectionPath(`/definitions/${definitionUuid}/runs`, companySlug, pageSize),
      undefined,
      options,
    );
  }

  listOutcomes(companySlug: string, definitionUuid: string, pageSize?: number, options?: RequestOptions) {
    return this.request<PredictionOutcomeSummary[]>(
      "GET",
      collectionPath(`/definitions/${definitionUuid}/outcomes`, companySlug, pageSize),
      undefined,
      options,
    );
  }

  getEvaluation(companySlug: string, definitionUuid: string, options?: RequestOptions) {
    return this.request<PredictionEvaluationSummary>(
      "GET",
      resourcePath(`/definitions/${definitionUuid}/evaluations`, companySlug),
      undefined,
      options,
    );
  }

  listThresholdEvents(companySlug: string, definitionUuid: string, pageSize?: number, options?: RequestOptions) {
    return this.request<PredictionThresholdEvent[]>(
      "GET",
      collectionPath(`/definitions/${definitionUuid}/threshold-events`, companySlug, pageSize),
      undefined,
      options,
    );
  }

  listSignalSources(companySlug: string, pageSize?: number, pageToken?: string, options?: RequestOptions) {
    return this.request<PredictionSignalSource[]>(
      "GET",
      collectionPath("/sources", companySlug, pageSize, pageToken),
      undefined,
      options,
    );
  }

  getSignalSource(companySlug: string, sourceUuid: string, options?: RequestOptions) {
    return this.request<PredictionSignalSource>(
      "GET",
      resourcePath(`/sources/${sourceUuid}`, companySlug),
      undefined,
      options,
    );
  }

  createSignalSource(companySlug: string, body: PredictionSignalSourceCreateRequest, options?: RequestOptions) {
    return this.request<PredictionSignalSource>("POST", collectionPath("/sources", companySlug), body, options);
  }

  activateSignalSource(companySlug: string, sourceUuid: string, options?: RequestOptions) {
    return this.request<void>("POST", resourcePath(`/sources/${sourceUuid}/activate`, companySlug), undefined, options);
  }

  archiveSignalSource(companySlug: string, sourceUuid: string, options?: RequestOptions) {
    return this.request<void>("POST", resourcePath(`/sources/${sourceUuid}/archive`, companySlug), undefined, options);
  }
}

function collectionPath(path: string, companySlug: string, pageSize?: number, pageToken?: string): string {
  const query = new URLSearchParams({ companySlug });
  if (pageSize !== undefined) query.set("pageSize", String(pageSize));
  if (pageToken !== undefined) query.set("pageToken", pageToken);
  return `/measurement/predictions${path}?${query.toString()}`;
}

function resourcePath(path: string, companySlug: string): string {
  const encodedPath = path
    .split("/")
    .map((segment, index) => (index === 0 ? segment : encodeURIComponent(segment)))
    .join("/");
  return collectionPath(encodedPath, companySlug);
}
